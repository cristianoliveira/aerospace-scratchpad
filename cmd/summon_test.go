package cmd

import (
	"fmt"
	"strings"
	"testing"

	aerospacecli "github.com/cristianoliveira/aerospace-ipc"
	"github.com/cristianoliveira/aerospace-scratchpad/internal/logger"
	mock_aerospace "github.com/cristianoliveira/aerospace-scratchpad/internal/mocks/aerospacecli"
	"github.com/cristianoliveira/aerospace-scratchpad/internal/stderr"
	"github.com/cristianoliveira/aerospace-scratchpad/internal/testutils"
	"github.com/gkampitakis/go-snaps/snaps"
	"go.uber.org/mock/gomock"
)

func TestSummonCmd(t *testing.T) {
	logger.SetDefaultLogger(&logger.EmptyLogger{})
	stderr.SetBehavior(false)

	t.Run("successfully summons a window by pattern", func(t *testing.T) {
		command := "summon"
		args := []string{command, "Notepad"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
					{
						AppName:  "Finder",
						WindowID: 5678,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 5678,
			},
		}
		allWindows := testutils.ExtractAllWindows(tree)
		focusedWorkspace := &aerospacecli.Workspace{Workspace: "ws1"}
		windows := testutils.ExtractWindowsByName(tree, "Notepad")
		if len(windows) != 1 {
			t.Fatalf("Expected 1 Notepad window, got %d", len(windows))
		}
		notepadWindow := windows[0]

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			// New behavior: get focused workspace before filtering windows
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedWorkspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(notepadWindow.WindowID, focusedWorkspace.Workspace).
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				SetFocusByWindowID(notepadWindow.WindowID).
				Return(nil).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err != nil {
			t.Errorf("Expected no error, got %v", err)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("fails when pattern doesn't match any window", func(t *testing.T) {
		command := "summon"
		args := []string{command, "NonExistentApp"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
					{
						AppName:  "Finder",
						WindowID: 5678,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 5678,
			},
		}
		allWindows := testutils.ExtractAllWindows(tree)
		focusedWorkspace := &aerospacecli.Workspace{Workspace: "ws1"}

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedWorkspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error, got nil")
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("fails when getting all windows returns an error", func(t *testing.T) {
		command := "summon"
		args := []string{command, "test"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(&aerospacecli.Workspace{Workspace: "ws1"}, nil).
				Times(1),
			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(nil, fmt.Errorf("mocked_error")).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error, got nil")
		}

		if out != "" {
			t.Errorf("Expected empty output, got %s", out)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("fails when getting focused workspace returns an error", func(t *testing.T) {
		command := "summon"
		args := []string{command, "Notepad"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 1234,
			},
		}
		_ = testutils.ExtractAllWindows(tree)

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		// New behavior: focused workspace is fetched before listing windows
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(nil, fmt.Errorf("mocked_error")).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error, got nil")
		}

		if out != "" {
			t.Errorf("Expected empty output, got %s", out)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("fails when regex pattern is invalid", func(t *testing.T) {
		command := "summon"
		args := []string{command, "[invalid"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		focusedWorkspace := &aerospacecli.Workspace{Workspace: "ws1"}

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		aerospaceClient.EXPECT().
			GetFocusedWorkspace().
			Return(focusedWorkspace, nil)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error, got nil")
		}

		if out != "" {
			t.Errorf("Expected empty output, got %s", out)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("fails when moving window to workspace returns an error", func(t *testing.T) {
		command := "summon"
		args := []string{command, "Notepad"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 1234,
			},
		}
		allWindows := testutils.ExtractAllWindows(tree)
		focusedWorkspace := &aerospacecli.Workspace{Workspace: "ws1"}
		windows := testutils.ExtractWindowsByName(tree, "Notepad")
		if len(windows) != 1 {
			t.Fatalf("Expected 1 Notepad window, got %d", len(windows))
		}
		notepadWindow := windows[0]

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedWorkspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(notepadWindow.WindowID, focusedWorkspace.Workspace).
				Return(fmt.Errorf("mocked_move_error")).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error, got nil")
		}

		if out != "" {
			t.Errorf("Expected empty output, got %s", out)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("fails when setting focus returns an error", func(t *testing.T) {
		command := "summon"
		args := []string{command, "Notepad"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 1234,
			},
		}
		allWindows := testutils.ExtractAllWindows(tree)
		focusedWorkspace := &aerospacecli.Workspace{Workspace: "ws1"}
		windows := testutils.ExtractWindowsByName(tree, "Notepad")
		if len(windows) != 1 {
			t.Fatalf("Expected 1 Notepad window, got %d", len(windows))
		}
		notepadWindow := windows[0]

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedWorkspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(notepadWindow.WindowID, focusedWorkspace.Workspace).
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				SetFocusByWindowID(notepadWindow.WindowID).
				Return(fmt.Errorf("mocked_focus_error")).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error, got nil")
		}

		if out != "" {
			t.Errorf("Expected empty output, got %s", out)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("summons multiple windows matching the pattern", func(t *testing.T) {
		command := "summon"
		args := []string{command, ".*(Notepad|TextEdit).*"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
					{
						AppName:  "TextEdit",
						WindowID: 5678,
					},
					{
						AppName:  "Finder",
						WindowID: 9012,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 9012,
			},
		}
		allWindows := testutils.ExtractAllWindows(tree)
		focusedWorkspace := &aerospacecli.Workspace{Workspace: "ws1"}
		windows := testutils.ExtractWindowsByName(tree, ".*(Notepad|TextEdit).*")
		if len(windows) != 2 {
			t.Fatalf("Expected 2 windows matching pattern, got %d", len(windows))
		}

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedWorkspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(windows[0].WindowID, focusedWorkspace.Workspace).
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				SetFocusByWindowID(windows[0].WindowID).
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(windows[1].WindowID, focusedWorkspace.Workspace).
				Return(nil).
				Times(1),

			aerospaceClient.EXPECT().
				SetFocusByWindowID(windows[1].WindowID).
				Return(nil).
				Times(1),
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err != nil {
			t.Errorf("Expected no error, got %v", err)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("handles empty pattern gracefully", func(t *testing.T) {
		command := "summon"
		args := []string{command, ""}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error for empty pattern, got nil")
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("handles whitespace-only pattern gracefully", func(t *testing.T) {
		command := "summon"
		args := []string{command, "   "}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err == nil {
			t.Errorf("Expected error for whitespace-only pattern, got nil")
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, cmdAsString, "Output", out, errorMessage)
	})

	t.Run("[dry-run] summons a window by pattern", func(t *testing.T) {
		command := "summon"
		args := []string{command, "Notepad", "--dry-run"}

		ctrl := gomock.NewController(t)
		defer ctrl.Finish()

		tree := []testutils.AeroSpaceTree{
			{
				Windows: []aerospacecli.Window{
					{
						AppName:  "Notepad",
						WindowID: 1234,
					},
					{
						AppName:  "Finder",
						WindowID: 5678,
					},
				},
				Workspace: &aerospacecli.Workspace{
					Workspace: "ws1",
				},
				FocusedWindowId: 5678,
			},
		}
		allWindows := testutils.ExtractAllWindows(tree)
		focusedWorkspace := &aerospacecli.Workspace{Workspace: "ws1"}
		windows := testutils.ExtractWindowsByName(tree, "Notepad")
		if len(windows) != 1 {
			t.Fatalf("Expected 1 Notepad window, got %d", len(windows))
		}
		notepadWindow := windows[0]

		aerospaceClient := mock_aerospace.NewMockAeroSpaceClient(ctrl)
		gomock.InOrder(
			aerospaceClient.EXPECT().
				GetFocusedWorkspace().
				Return(focusedWorkspace, nil).
				Times(1),

			aerospaceClient.EXPECT().
				GetAllWindows().
				Return(allWindows, nil).
				Times(1),

			// In dry-run mode, the aerospace client wrapper intercepts these calls
			// and prints debug messages instead of calling the actual methods
			aerospaceClient.EXPECT().
				MoveWindowToWorkspace(notepadWindow.WindowID, focusedWorkspace.Workspace).
				Return(nil).
				Times(0), // DO NOT RUN in dry-run mode

			aerospaceClient.EXPECT().
				SetFocusByWindowID(notepadWindow.WindowID).
				Return(nil).
				Times(0), // DO NOT RUN in dry-run mode
		)

		cmd := RootCmd(aerospaceClient)
		out, err := testutils.CmdExecute(cmd, args...)
		if err != nil {
			t.Errorf("Expected no error, got %v", err)
		}

		cmdAsString := "aerospace-scratchpad " + strings.Join(args, " ") + "\n"
		errorMessage := fmt.Sprintf("Error\n %+v", err)
		snaps.MatchSnapshot(t, tree, cmdAsString, "Output", out, errorMessage)
	})
}
