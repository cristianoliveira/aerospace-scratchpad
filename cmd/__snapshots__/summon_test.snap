
[TestSummonCmd/successfully_summons_a_window_by_pattern - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowId: 5678,
    },
}
aerospace-scratchpad summon Notepad

Output
Window '1234 | Notepad ' is moved to workspace 'ws1'

Error
 <nil>
---

[TestSummonCmd/fails_when_pattern_doesn't_match_any_window - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowId: 5678,
    },
}
aerospace-scratchpad summon NonExistentApp

Output

Error
 Error: no windows matched the pattern 'NonExistentApp'

---

[TestSummonCmd/fails_when_getting_all_windows_returns_an_error - 1]
aerospace-scratchpad summon test

Output

Error
 Error: unable to get windows: mocked_error

---

[TestSummonCmd/fails_when_getting_focused_workspace_returns_an_error - 1]
aerospace-scratchpad summon Notepad

Output

Error
 Error: unable to get focused workspace

---

[TestSummonCmd/fails_when_regex_pattern_is_invalid - 1]
aerospace-scratchpad summon [invalid

Output

Error
 Error: invalid app-name-pattern, error parsing regexp: missing closing ]: `[invalid`

---

[TestSummonCmd/fails_when_moving_window_to_workspace_returns_an_error - 1]
aerospace-scratchpad summon Notepad

Output

Error
 Error: unable to move window '1234 | Notepad ' to workspace 'ws1': mocked_move_error

---

[TestSummonCmd/fails_when_setting_focus_returns_an_error - 1]
aerospace-scratchpad summon Notepad

Output

Error
 Error: unable to set focus to window '1234 | Notepad ': mocked_focus_error

---

[TestSummonCmd/summons_multiple_windows_matching_the_pattern - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"TextEdit", AppBundleID:"", Workspace:""},
            {WindowID:9012, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowId: 9012,
    },
}
aerospace-scratchpad summon .*(Notepad|TextEdit).*

Output
Window '1234 | Notepad ' is moved to workspace 'ws1'
Window '5678 | TextEdit ' is moved to workspace 'ws1'

Error
 <nil>
---

[TestSummonCmd/handles_empty_pattern_gracefully - 1]
aerospace-scratchpad summon 

Output

Error
 argument at position 0 is empty or whitespace
---

[TestSummonCmd/handles_whitespace-only_pattern_gracefully - 1]
aerospace-scratchpad summon    

Output

Error
 argument at position 0 is empty or whitespace
---

[TestSummonCmd/[dry-run]_summons_a_window_by_pattern - 1]
[]testutils.AeroSpaceTree{
    {
        Windows: {
            {WindowID:1234, WindowTitle:"", AppName:"Notepad", AppBundleID:"", Workspace:""},
            {WindowID:5678, WindowTitle:"", AppName:"Finder", AppBundleID:"", Workspace:""},
        },
        Workspace:       &aerospace.Workspace{Workspace:"ws1"},
        FocusedWindowId: 5678,
    },
}
aerospace-scratchpad summon Notepad --dry-run

Output
[dry-run] MoveWindowToWorkspace(windowID=1234, workspace=ws1)
[dry-run] SetFocusByWindowID(1234)
Window '1234 | Notepad ' is moved to workspace 'ws1'

Error
 <nil>
---
